tests/upgrade/test_upgrade_from.sh

```diff
@@ -0,0 +1,228 @@
+}
+
+writeMsg() {
+    printf "\n\n****************\n\n${1}\n\n****************\n\n"
+
+}
+
+waitForIngress() {
+    sleep_time_sec=10
+    INGRESS_HOST=""
+    while [ -z ${INGRESS_HOST} ]; do
+        echo "Waiting for ingress-gateway addr..."
+        INGRESS_HOST=$(kubectl -n istio-system get service istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+        INGRESS_PORT=$(kubectl -n istio-system get service istio-ingressgateway -o jsonpath='{.spec.ports[?(@.name=="http2")].port}')
+        INGRESS_ADDR=${INGRESS_HOST}:${INGRESS_PORT}
+        sleep ${sleep_time_sec}
+        (( sleep_time_sofar += $sleep_time_sec ))
+        if (( sleep_time_sofar > 300 )); then
+            echo "Timed out waiting for ingress-gateway address"
+            exit 1
+        fi
+    done
+    echo "Got ingress-gateway addr: ${INGRESS_ADDR}"
+}
+
+
+waitForPodsReady() {
+    sleep_time_sec=10
+    while :; do
+        echo "Waiting for pods to be ready in ${1}..."
+        pods_str=$(kubectl -n ${1} get pods | tail -n +2 )
+        arr=()
+        while read -r line; do
+           arr+=("$line")
+        done <<< "$pods_str"
+
+        ready="true"
+        for line in "${arr[@]}"; do
+            if [[ ${line} != *"Running"* ]]; then
+                ready="false"
+            fi
+        done
+        if [  "${ready}" = "true" ]; then
+            echo "All pods ready."
+            return
+        fi
+
+        sleep ${sleep_time_sec}
+        (( sleep_time_sofar += $sleep_time_sec ))
+        if (( sleep_time_sofar > 300 )); then
+            echo "Timed out waiting for pods to be ready."
+            echo ${pods_str}
+            exit 1
+        fi
+    done
+}
+
+checkEchosrv() {
+    sleep_time_sec=10
+    while :; do
+        resp=$( curl -o /dev/null -s -w "%{http_code}\n" -HHost:echosrv.${TEST_NAMESPACE}.svc.cluster.local http://${INGRESS_ADDR}/echo )
+        if [[ "${resp}" = *"200"* ]]; then
+            echo "Got correct response from echosrv."
+            return
+        else
+            echo "Got bad echosrv response: ${resp}"
+        fi
+
+        sleep ${sleep_time_sec}
+        (( sleep_time_sofar += $sleep_time_sec ))
+        if (( sleep_time_sofar > 300 )); then
+            echo "Timed out waiting for 200 from echosrv."
+            exit 1
+        fi
+    done
+}
+
+git fetch upstream --tags --prune
+git fetch origin --tags --prune
+
+kubectl create namespace istio-system
+kubectl create namespace ${TEST_NAMESPACE}
+kubectl label namespace ${TEST_NAMESPACE} istio-injection=enabled
+
+pushd ${ISTIO_ROOT}
+
+installIstioSystemAtTag ${FROM_HUB} ${FROM_TAG}
+waitForIngress
+waitForPodsReady istio-system
+
+installTest
+waitForPodsReady ${TEST_NAMESPACE}
+sleep 20
+checkEchosrv
+
+sendInternalRequestTraffic
+sendExternalRequestTraffic ${INGRESS_ADDR}
+sleep 20
+installIstioSystemAtVersionHelmTemplate ${HUB} ${TAG}
+waitForPodsReady istio-system
+sleep 60
+restartDataPlane echosrv-deployment-v1
+sleep 100
```

Comment:
Could we not do a wait for the pods rather than sleeping?

Reply:
In this case no because it's an async operation and all pods are often ready due to restart delay controlled by minRestartSeconds. Added comment.